include/seastar/util/process.hh

```diff
@@ -0,0 +1,112 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (C) 2022 Kefu Chai ( tchaikov@gmail.com )
+ */
+
+#pragma once
+
+#include <sys/types.h>
+#include <algorithm>
+#include <filesystem>
+#include <initializer_list>
+#include <iterator>
+#include <string_view>
+#include <utility>
+#include <variant>
+#include <vector>
+#include <fmt/format.h>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/sstring.hh>
+
+namespace seastar::experimental {
+
+/// The optional parameters for spawning a subprocess
+///
+/// \note see \c execve(2) for more details on \c argv and \c env.
+struct spawn_parameters {
+    /// The arguments passed to the program
+    std::vector<sstring> argv;
+    /// The environment variables for the program
+    std::vector<sstring> env;
+    spawn_parameters(std::initializer_list<std::string_view> argv)
+        : argv(argv.begin(), argv.end()) {}
+    spawn_parameters(std::initializer_list<std::string_view> argv,
+                     std::initializer_list<std::pair<std::string_view, std::string_view>> envs)
+        : argv(argv.begin(), argv.end())
+    {
+        std::transform(envs.begin(), envs.end(), std::inserter(env, env.end()),
+                       [](const auto& e) {
+                           return fmt::format("{}={}", e.first, e.second);
+                       });
+    }
+};
+/// Interact with a spawned subprocess
+///
+/// \note the spawned subprocess should always be \c wait()'ed. Otherwise,
+/// the Seastar application spawning the subprocess will leave us with
+/// one ore more zombie subprocesses after it exists.
+class process {
+    struct create_tag {};
+public:
+    process(create_tag, pid_t pid, int stdin, int stdout, int stderr);
+    /// Spawn a subprocess using \c posix_spawn(3)
+    ///
+    /// \param pathname the full path to the executable
+    /// \param params parameters for spawning the subprocess
+    ///
+    /// \returns a \c process instance representing the spawned subprocess
+    static future<process> spawn(const std::filesystem::path& pathname,
+                                 spawn_parameters params);
+    /// Spawn a subprocess using \c posix_spawn(3)
+    ///
+    /// \param pathname the full path to the executable
+    ///
+    /// \returns a \c process instance representing the spawned subprocess
+    static future<process> spawn(const std::filesystem::path& pathname);
+    /// Return an writable stream which provides input from the child process
+    output_stream<char>& stdin() {
+        return _stdin;
+    }
+    /// Return an writable stream which provides stdout output from the child process
+    input_stream<char>& stdout() {
+        return _stdout;
+    }
+    /// Return an writable stream which provides stderr output from the child process
+    input_stream<char>& stderr() {
+        return _stderr;
+    }
+    struct wait_exited {
```

Comment:
Should those move away stdin/stdout? Like connected_socket? I guess it's better to keep the code as is.

Reply:
instead of constructing these streams in the ctor, the latest revision construct them right in the accessors.